src/yaml/from_yaml.cr

```diff
@@ -138,47 +166,83 @@ def NamedTuple.new(pull : YAML::PullParser)
   {% end %}
 end
 
-def Enum.new(pull : YAML::PullParser)
-  string = pull.read_scalar
+def Enum.new(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
+  unless node.is_a?(YAML::Nodes::Scalar)
+    node.raise "Expected scalar, not #{node.class}"
+  end
+
+  string = node.value
   if value = string.to_i64?
     from_value(value)
   else
     parse(string)
   end
 end
 
-def Union.new(pull : YAML::PullParser)
-  location = pull.location
-  string = pull.read_raw
+def Union.new(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
+  if node.is_a?(YAML::Nodes::Alias)
+    {% for type in T %}
+      {% if type < ::Reference %}
```

Comment:
can you describe this trick? looks like you want `!type.is_a?(Reference)`

Reply:
It means "if the type inherits from Reference". If so, we can check for an alias.

Reply:
looks like Crystal need explicit method named `.reference?`?

Reply:
it returns `true` if _type_ inherits from _::Reference_.

Reply:
Yes, but note that this only works in macros. In runtime this isn't possible (yet)